### PR TITLE
Imporve performance when generating spec with external dependencies

### PR DESCRIPTION
**Pull request by pytimer**
**Describe the PR**

Directly use `go list` command instead of `depth`, because `depth` use go standard library `build.Import`  and call the `build.Import` too many times, it will spent too many times when the large dependencies.

Because `swag` only needs to know which files are dependent, and does not need to get the level of dependencies, all dependencies can be directly through `go list` command, and `swag` not need to go through deep loop again, but if use `depth`, `swag` still need to loop too many times, so i refactor the swag get ast files code.

**Relation issue**

Fixes: https://github.com/swaggo/swag/issues/1032

**Additional context**

Added the `--parseGoList` flag, the default value is true default. If someone does not want to enable it, it can be set to false when init, if it's set to false, `swag` still use `depth` for dependency parser.

### CODE REVIEW

**Comment by ubogdan** [comment]
Unfortunately, I will have a busy week.

@easonlin404 can you please assist with a CR here?

**Comment by magandrez** [comment]
@ubogdan @easonlin404 Is there any way to contribute to this change? please, let know.

**Comment by ubogdan** [comment]
This implementation introduces a dependency over an external binary, the go binary, that comes with some new challenges :
1. The output format may change over time.
2. You may get slightly different output on different go versions.
3. You may run the generator in an environment where the GO binary it's not in PATH.
4. You may run the generator in an environment where the GO binary is not the same as your one to build your project.

I prefer to see this done programmatically by importing a go library that handlers go.mod and calling the specific methods.

The solution proposed by @pytimer may look like a quick fix at this time. Once the feature gets merged to master, the community will start using it, more issues will arrive, and code complexity will grow.

**Comment by pytimer** [comment]
@ubogdan `swag` currently already depends on the go binary, now if the env has not go binary, `swag init` failed, because init first call `go list` to get pkg name, the code is [getPkgName](https://github.com/swaggo/swag/blob/master/parser.go#L330).
`depth` call `build.Default`, this code also call the go binary file, if no go binary, it will return `ErrRootPkgNotResolved`, this pr does not newly introduce the go binary.

However, the two points 2 and 4 above @ubogdan said do need to be pay attention to. I don't know much about the output of multiple versions of go list. I only tested it on versions of go 1.16 and 1.17. 
I will try to understand the output format of go list in different versions. If anyone else is familiar with it, please point it out.

**Comment by ubogdan** [comment]
> @ubogdan swag currently already depends on the go binary, 
Please don't return this as an excuse. 
I explained why I wouldn't say I like the idea, but didn't refuse the implementation. 
It's more simple for me to Approve and Merge and after that, I can tag you in every issue where a community member is complaining something is wrong.

**Comment by g1eny0ung** [comment]
Any updates? After my local testing, this will greatly improve the generation speed.

**Comment by ubogdan** [comment]
I will consider this total silence as an aproval. 
@pytimer I will go forward with a code review in the next few days, please confirm you are still interested to finish this PR.

**Comment by pytimer** [comment]
Thanks.
I read some docs about not use binary execute command to do it recently, but unfortunately i don't find.

I try golang.org/x/tools/go/packages.Load, but it not works well.

Perhaps use `go list` now and find the better ways to optimize in later version, what's your recommend?

**Comment by ubogdan** [comment]
We will go with the current implementation. Please confirm you still intend to finish this PR so I can go with a CR.

**Comment by pytimer** [comment]
> We will go with the current implementation. Please confirm you still intend to finish this PR so I can go with a CR. 

Yes，please tell me if need I do.

File: golist.go
Comment (ubogdan) [inline golist.go:16]: `goArgs` is used only once
```go
	cmd := exec.CommandContext(ctx, "go", append([]string{"list", "-json", "-e"}, args...)...)
```

File: golist.go
Comment (ubogdan) [inline golist.go:38]: Same here ...
```go
                err = dec.Decode(&pkg);
		if  err != nil {
```

File: golist.go
Comment (ubogdan) [inline golist.go:43]: Same here
```go
        err = cmd.Wait(); 
	if err != nil {
		return nil, err
```

File: golist.go
Comment (ubogdan) [inline golist.go:59]: var err error

File: golist.go
Comment (ubogdan) [inline golist.go:70]: Same here
```go
                err = parser.parseFile(pkg.ImportPath, filepath.Join(srcDir, pkg.CFiles[i]), nil); 
		if err != nil {
			return err

File: parser.go
Comment (ubogdan) [inline parser.go:146]: Make this private, will be updated by 

```go
// ParseUsingGoList --fill--compent---please.
func ParseUsingGoList(Enabled bool) func(*Parser) {
	return func(p *Parser) {
		p.parseGoList = Enabled
	}
}
```

**Review by ubogdan — CHANGES_REQUESTED** [review]
See comments ... Also include the usage example in the README.md

**Comment by pytimer** [comment]
@ubogdan Update the code, you can review it if you have time, thanks.

**Review by ubogdan — APPROVED** [review]
LGTM

**Comment by ubogdan** [comment]
> @ubogdan Update the code, you can review it if you have time, thanks.
> 
> Should i squash all commits after review or now?

We are using  "Squash and Merge" for all the contributions.

File: golist.go
Comment (ubogdan) [inline golist.go:59]: Please cover this by calling the function with a build.Package instance.

**Comment by ubogdan** [comment]
Please improve code coverage so we can merge it.

File: golist.go
Comment (pytimer) [reply golist.go:59]: ok, i will add some testing on this Saturday, i don't have the bandwidth these days, sorry.

**Review by akhmadKurniawan — COMMENTED** [review]
ok

**Comment by sdghchj** [comment]
> I expect to see at least one opinion from @easonlin404 or @sdghchj regarding this matter before going forward to do a real CR and approve it.

As far as I see, `go list` will find out all its dependent packages, I am not sure whether it's more efficient in a huge project or not.

**Comment by magandrez** [comment]
Any ETA to land this? @ubogdan

**Comment by ubogdan** [comment]
@magandrez, I'm waiting for @pytimer to improve code coverage so we can merge it.

**Comment by g1eny0ung** [comment]
@ubogdan I wonder if it is possible to draft a new release after merging this PR? I've been looking forward to this improvement for a long time. Very grateful to everyone for this.

**Comment by pytimer** [comment]
@ubogdan How can i rerun github Actions and codecov?

**Comment by ubogdan** [comment]
Good job. Please resolve conflicts.

**Comment by pytimer** [comment]
@ubogdan solve conflicts, please review it, thanks.

**Comment by ubogdan** [comment]
Looks like code coverage is still failing :(

**Comment by pytimer** [comment]
Maybe i merge master branch code, add test code in the latest commit.

**Comment by pytimer** [comment]
Please rerun test codecov, thanks.

**Comment by pytimer** [comment]
@ubogdan Can you tell me how to fix codecov/patch, i don't understand where the problem, thanks.

**Comment by ubogdan** [comment]
Maybe you need to write unit tests this parts of the code: https://github.com/swaggo/swag/pull/1108/checks?check_run_id=6346809522?

**Review by ubogdan — APPROVED** [review]
LGTM

**Comment by ubogdan** [comment]
> y ETA regarding the required changes

We will try to include this in the next release. I was hoping we would get better code coverage.

**Comment by g1eny0ung** [comment]
@ubogdan Can we release a version? It's been a long time since this PR was merged. 😢

**Comment by ubogdan** [comment]
released in v1.8.3

**Comment by g1eny0ung** [comment]
> released in v1.8.3

Many thanks!

**Pull request merged**
